File: C++/problem0870.cpp
#include <iostream>
#include <iomanip>
#include <vector>
#include <map>
#include <set>
#include <algorithm>
#include <cmath>
#include <ctime>
#include <cstdlib>
#include <windows.h>
#include <string>
#include <numeric>

using namespace std;

class Solution {
public:
    vector<int> advantageCount(vector<int>& A, vector<int>& B) {
        vector<int> originalB = B;
        sort(A.begin(), A.end());
        sort(B.begin(), B.end());
        vector<int>::iterator it;
        int k = 0;
        vector<int> tmp;
        map<int, vector<int>> d;
        for (it = A.begin(); it < A.end(); ++it)
        {
            if (*it > B[k])
            {
                d[B[k]].push_back(*it);
                k++;
            }
            else
            {
                tmp.push_back(*it);
                // cout << *it << endl;
            }
        }
        vector<int> ans;
        for (it = originalB.begin(); it < originalB.end(); ++it)
        {
            if (!d[*it].empty())
            {
                // cout << d[*it][d[*it].size()-1] << endl;
                ans.push_back(d[*it][d[*it].size()-1]);
                d[*it].pop_back();
            }
            else
            {
                // cout << tmp[tmp.size()] << endl;
                ans.push_back(tmp[tmp.size()-1]);
                tmp.pop_back();
            }
        }
        return ans;
    }
};

int main()
{
    // vector<int> A {2,7,11,15}, B {1,10,4,11}, ans;
    vector<int> A {12,24,8,32}, B {13,25,32,11}, ans;
    Solution solu;
    ans = solu.advantageCount(A, B);
    for (int k = 0; k < ans.size(); ++k)
    {
        cout << ans[k] << '\t';
    }
    cout << endl;
    return 0;
}

File: Python/problem0870.py
from typing import List

class Solution:
    def advantageCount(self, A: List[int], B: List[int]) -> List[int]:
        # if not (A and B):
        #     return A
        # d = {}
        # for k in range(len(B)):
        #     d[B[k]] = [k]
        # A.sort()
        # B.sort()
        # for k in range(len(B)):
        #     d[B[k]].append(k)
        # ans = []
        # tmp = []
        # p, q = 0, 0
        # while p < len(A) and q < len(B):
        #     if A[p] > B[q]:
        #         ans.append(A[p])
        #         p += 1
        #         q += 1
        #     else:
        #         tmp.append(A[p])
        #         p += 1
        # ans = ans + tmp
        # s = set()
        # for value in d.values():
        #     tmp = value
        #     s.add(tmp[0])
        #     if tmp[1] not in s:
        #         ans[tmp[0]], ans[tmp[1]] = ans[tmp[1]], ans[tmp[0]]
        # return ans
        
        # if not (A and B):
        #     return A
        # k = 0
        # sA, sB = sorted(A), sorted(B)
        # d, D = {}, {}
        # for a in sA:
        #     if a > sB[k]:
        #         d[sB[k]] = a
        #         k += 1
        #     else:
        #         D[sB[k]] = a
        # print(d)
        # print(D)
        # ans = []
        # for b in B:
        #     try:
        #         ans.append(d[b])
        #     except:
        #         ans.append(D[b])
        # return ans

        if not (A and B):
            return A
        sortedA, sortedB = sorted(A), sorted(B)
        d = {b:[] for b in B}
        tmp = []
        k = 0
        for a in sortedA:
            if a > sortedB[k]:
                d[sortedB[k]].append(a)
                k += 1
            else:
                tmp.append(a)
        return [d[b].pop() if d[b] else tmp.pop() for b in B]


solu = Solution()
A, B = [2,7,11,15], [1,10,4,11]
# A, B = [12,24,8,32], [13,25,32,11]
print(solu.advantageCount(A, B))

